service-api/module/Application/src/DWP/DwpApi/DwpApiService.php

```diff
@@ -0,0 +1,248 @@
+                new DetailsRequestDTO($citizenResponseDTO->id()),
+                $nino
+            );
+
+            return $this->compareRecords($caseData, $detailsResponseDTO, $citizenResponseDTO, $nino);
+        } catch (\Exception $exception) {
+            $this->logger->error('DwpApiException: ' . $exception->getMessage());
+            throw new DwpApiException($exception->getMessage());
+        }
+    }
+
+    public function compareRecords(
+        CaseData $caseData,
+        DetailsResponseDTO $detailsResponseDTO,
+        CitizenResponseDTO $citizenResponseDTO,
+        string $nino
+    ): bool {
+        if ($caseData->idMethodIncludingNation) {
+            /** @psalm-suppress PossiblyNullArgument */
+            $submittedNino = strtoupper(
+                preg_replace(
+                    '/(\s+)|(-)/',
+                    '',
+                    $nino
+                )
+            );
+
+            /** @psalm-suppress PossiblyNullArgument */
+            $returnedNino = strtoupper(
+                preg_replace(
+                    '/(\s+)|(-)/',
+                    '',
+                    $detailsResponseDTO->nino()
+                )
+            );
+
+            if ($citizenResponseDTO->type() !== 'MatchResult') {
+                $this->logger->info("Match Result missing from response.");
+                return false;
+            }
+
+            if ($submittedNino !== $returnedNino) {
+                $this->logger->info("Submitted NINO does not match NINO in details response.");
+                return false;
+            }
+
+            return true;
+        } else {
+            throw new DwpApiException('National Insurance Number not set.');
+        }
+    }
+
+    /**
+     * @throws GuzzleException
+     * @throws DwpApiException
+     */
+    public function makeCitizenMatchRequest(
+        CitizenRequestDTO $citizenRequestDTO,
+    ): CitizenResponseDTO {
+        $this->authCount++;
+        $responseArray = [];
+        try {
+            $postBody = $this->constructCitizenRequestBody($citizenRequestDTO);
+            $this->logger->info('MATCH_POSTBODY: ', $postBody);
+
+            $response = $this->guzzleClient->request(
+                'POST',
+                $this->matchPath,
+                [
+                    'headers' => $this->makeHeaders(),
+                    'json' => $postBody
+                ]
+            );
+            $responseArray = json_decode($response->getBody()->getContents(), true);
+            $this->logger->info("DWP_MATCH_RESPONSE: " . json_encode($responseArray, JSON_THROW_ON_ERROR));
+        } catch (ClientException $clientException) {
+            if (
+                $clientException->getResponse()->getStatusCode() == Response::STATUS_CODE_401 &&
+                $this->authCount < 2
+            ) {
+                $this->authApiService->authenticate();
+                $this->makeCitizenMatchRequest($citizenRequestDTO);
```

Comment:
Don't we need to return the value here?
```
return $this->makeCitizenMatchRequest($citizenRequestDTO);
```

Reply:
No, it's in the responseDto at the end of the function

Reply:
I think you still need to set `$responseArray`, otherwise the value is never captured from the second request
```suggestion
                $responseArray = $this->makeCitizenMatchRequest($citizenRequestDTO);
```